web/src/utils/refetching.ts

```diff
@@ -0,0 +1,33 @@
+import type { QueryClient } from '@tanstack/react-query';
+import { ONE_MINUTE, QUERY_KEYS } from 'api/helpers';
+import { TimeAverages } from './constants';
+
+/**
+ * Refetches data when the hour changes to show fresh data.
+ * Checks every minute if we have passed the hour and invalidates hourly queries
+ * if so. Invalidating the query will cause it to refetch if it is actively in use by a mounted
+ * component (e.g. the map or left panel for a specific zone), otherwise it will refetch the next
+ * time the component is mounted (e.g. when going back to a zone previously viewed)
+ */
+export function refetchDataOnHourChange(queryClient: QueryClient) {
+  let startHour = new Date().getMinutes();
+  setInterval(() => {
+    const currentHour = new Date().getMinutes();
+    if (startHour !== currentHour) {
+      console.info(`Refetching data for new hour: ${currentHour}`);
+
+      // Invalidate hourly state query
+      queryClient.invalidateQueries({
+        queryKey: [QUERY_KEYS.STATE, { aggregate: TimeAverages.HOURLY }],
+      });
+
+      // Invalidate hourly zone queries - this matches all zone queries
+      queryClient.invalidateQueries({
+        queryKey: [QUERY_KEYS.ZONE, { aggregate: TimeAverages.HOURLY }],
+      });
+
+      // Reset the start hour
+      startHour = currentHour;
+    }
+  }, ONE_MINUTE);
+}
```

Comment:
```suggestion
export function refetchDataOnHourChange(queryClient: QueryClient) {
  let startHour = new Date().getUTCHours();
  setInterval(() => {
    const currentHour = new Date().getUTCHours();
    if (startHour !== currentHour) {
      console.info(`Refetching data for new hour: ${currentHour}`);

      // Invalidate hourly state query
      queryClient.invalidateQueries({
        queryKey: [QUERY_KEYS.STATE, { aggregate: TimeAverages.HOURLY }],
      });

      // Invalidate hourly zone queries - this matches all zone queries
      queryClient.invalidateQueries({
        queryKey: [QUERY_KEYS.ZONE, { aggregate: TimeAverages.HOURLY }],
      });

      // Reset the start hour
      startHour = currentHour;
    }
  }, ONE_MINUTE);
}
```

This should probably use hours instead of minutes right?
(UTC hours prevents timezone weirdness from interfering with this)

Reply:
I'm thinking it might be better to set this as a timeout with the following value instead though to avoid checking it all the time?
```js
  const currentDate = new Date();
  const refetchIn =
    (60 * 60 * 1000) -
    (currentDate.getUTCMinutes() * 60 * 1000 +
      currentDate.getUTCSeconds() * 1000 +
      currentDate.getUTCMilliseconds());
  setTimeout(() => console.log('Refetch Now'), refetchIn);
```
This is just a late night thought though, I don't really know what is best or more performant.

Reply:
Alternatively we could use the refetchIn variable that Viktor has suggested here as the refetchInterval.

```
const getTimeToNextHour = () => {
  const currentDate = new Date();
  const refetchIn =
    60 * 60 * 1000 -
    (currentDate.getUTCMinutes() * 60 * 1000 +
      currentDate.getUTCSeconds() * 1000 +
      currentDate.getUTCMilliseconds());

  return refetchIn;
};

const useGetState = (): UseQueryResult<GridState> => {
  const [timeAverage] = useAtom(timeAverageAtom);
  const timeToNextHour = getTimeToNextHour();
  const [refetchInterval, setRefetchInterval] = useState(timeToNextHour);
  return useQuery<GridState>(
    [QUERY_KEYS.STATE, timeAverage],
    async () => getState(timeAverage),
    {
      refetchInterval,
      staleTime: ONE_HOUR,
      refetchOnWindowFocus: true,
      onSuccess: () => setRefetchInterval(getTimeToNextHour()),
    }
  );
};

export default useGetState;
```

Reply:
Whoops yes, just changed it to minutes to make the preview video 😅 

I thought about using refetchInterval too, but if I visit the app at 4:55pm the refetchInterval would be 5 minutes,  and after the hour change we don't want it to refetch that often. So I don't think calculating the "time until next hour" just once is going to work :)

Reply:
That's why I added this line :D

```onSuccess: () => setRefetchInterval(getTimeToNextHour()),```

Reply:
Ah sorry, didn't see the code-block was scrollable 😂 Will test it!

Reply:
This function has been deprecated though, I liked that this wouldn't reinvent the wheel but maybe it's best to ignore this approach.

Reply:
Where do you see that `onSuccess` is deprecated?

Reply:
https://tanstack.com/query/latest/docs/react/reference/useQuery
![image](https://github.com/electricitymaps/electricitymaps-contrib/assets/18622768/f51e0c01-941b-43f1-9f4a-3f4bfd0fbf79)

Reply:
Ah thanks, turns out they even did a blog post about it: https://tkdodo.eu/blog/breaking-react-querys-api-on-purpose

As an alternative, we could do something like this using a useEffect hook, but it also gets quite messy and I'm unsure it if will cause bugs since we use the getZone hook in multiple places 🤔 



![Screenshot 2023-06-26 at 11 20 10](https://github.com/electricitymaps/electricitymaps-contrib/assets/3296643/c65aad74-d579-4aab-bd09-252764d40357)

![Screenshot 2023-06-26 at 11 21 41](https://github.com/electricitymaps/electricitymaps-contrib/assets/3296643/2117383a-120f-4bb8-97cc-023ed77185ce)

Reply:
I messaged Dominik on twitter about our use case and the deprecated function, this was his reply:
![image](https://github.com/electricitymaps/electricitymaps-contrib/assets/18622768/0aeb5112-4ef6-4525-9439-36c23334f5ef)